[Driver] [GPIO] pca_series gpio expander driver

CkovMk opened the pull request:
## Scope

I was trying to write a new driver for PCAL6524, a i2c gpio expander. Going thourgh existing drivers, there are numbers of driver for different PCA(L) series devices. They share similiar register layout and control logic. So I intends to unify these drivers for PCA(L)xxxx series i2c gpio expanders.
If this driver gets merged, it should retire `pca95xx` , `pca953x` and `pcal64xxa` drivers.

## Method

By referencing exsisting `gpio_pca95xx` and `pcal64xxa` driver, I have drafted these code to show my intended implementation. As shown in code, different capablilties can be defined:

- basic functions
  this applies to all compatiable devices.
- "latch" and drive strength
  this is what "L" in PCALxxxx means. allows to "latch" input until read and configure drive strength to one of 4 levels.
- pull up/down
  pull enable and pull select
- interrupt mask
  able to mask interrupt on certain pins. most PCAL devices has this, as well as PCA957x.
- interrupt edge / clear
  seen on PCAL6524.
- output config
  seen on PCAL6524. note that although many devices is able to configure push-pull or open-drain on a per-port basis, PCAL6524 suports configure on a per-pin basis.

We can describe a compatible device using a set of feature flags and `ngpios` property. Based on that, it's possible to create separate register maps and api maps to drive different devices.

## Question

1. Is it necessary to maintain cache for all config registers? 
  In the original code, lots of registers are "cached" to speed-up the "read-modify-write" process. In my view, I/O expanders are not meant to be frequenty reconfigured. We also do not have tools like regmap in linux to automate the caching. Maintaining cache for devices with different port size can be complicated in coding. My intention is to only cache output values.
  update: The final implementation use Kconfig to control caching. I've tested the `GPIO_PCA_SERIES_CACHE_ALL=n` option on pcal6524 on `mimx93_evk_a55` board. The `GPIO_PCA_SERIES_CACHE_ALL=y` option remains to be fully tested.

2. Is there way to avoid creating multiple dt-bindings?
  I tried to avoid creating a lot of dt-bindings by put a "part_no" property under a single `nxp,pca_series.yaml` dt-binding.

Any suggestions are welcomed. 

Thanks

CkovMk commented:
This is a re-open of #63101 .

MaureenHelm commented:
@dleach02 @danieldegrasse please take a look

danieldegrasse commented on dts/bindings/gpio/nxp,pca_series.yaml:
```diff
@@ -0,0 +1,55 @@
+include: [gpio-controller.yaml, i2c-device.yaml]
+
+properties:
+  part_no:
```
I don't think this how we should support multiple devices with one driver. Zephyr drivers that support multiple compatibles typically do so via a pattern like so:

```
/* In this case, type would be a parameter indicating the part number of the device, based on compatible */
#define MY_DRV_INIT_MACRO(inst, type)

#define DT_DRV_COMPAT xxxx
DT_INST_FOREACH_STATUS_OKAY_VARGS(MY_DRV_INIT_MACRO, XXXX)

#undef DT_DRV_COMPAT
#define DT_DRV_COMPAT yyyy
DT_INST_FOREACH_STATUS_OKAY_VARGS(MY_DRV_INIT_MACRO, YYYY)
````

See the lp50xxx driver for an example of this method: https://github.com/zephyrproject-rtos/zephyr/blob/4c4ff203837bd17ffbcd5e5dc8bd90818dda74b4/drivers/led/lp50xx.c#L400

Another approach we could use would be something like what the ILI9XXX driver does: https://github.com/zephyrproject-rtos/zephyr/blob/4c4ff203837bd17ffbcd5e5dc8bd90818dda74b4/drivers/display/display_ili9xxx.c#L411

I think either approach would be fine- but we should avoid encoding part number like this

CkovMk replied:
use "DT_INST_FOREACH_STATUS_OKAY_VARGS" to pass part_no.

danieldegrasse commented on drivers/gpio/gpio_pca_series.c:
```diff
@@ -0,0 +1,2077 @@
+ * - Type 4: PCAL6524, PCAL6534
+ */
+
+enum gpio_pca_series_reg_type {				/** Type1 Type2 Type3 Type4 */
```
Convention in Zephyr is typically to capitalize all `enum` constants- this comment applies to the additional enums below as well

CkovMk replied:
renamed.

danieldegrasse commented on drivers/gpio/gpio_pca_series.c:
```diff
@@ -0,0 +1,2077 @@
+						   enum gpio_pca_series_reg_type reg_type)
+{
+	const struct gpio_pca_series_config *cfg = dev->config;
+	/* struct gpio_pca_series_data *data = dev->data; */
```
If code like this is not needed, please remove it rather than commenting it out

danieldegrasse commented on drivers/gpio/gpio_pca_series.c:
```diff
@@ -0,0 +1,2077 @@
+#endif
+};
+
+/*!
```
I'm not clear why doxygen style macros are being used here- I have no issue with doxygen style function documentation in drivers, but these kinds of macros are only really useful when doxygen will be parsing the file- and to my knowledge doxygen does not parse the driver files themselves- so creating a doxygen group here won't be useful

CkovMk replied:
Yeah, it is just used to remind me what these code does. We can remove these if confusing.

danieldegrasse replied:
Maybe instead we expand the comment? We aren't limited by doxygen syntax here, this commend could instead say "PCA GPIO register access functions"

CkovMk replied:
changed this to avoid creating doxygen groups. it seems they are causing ci doc build failure.

danieldegrasse commented on drivers/gpio/gpio_pca_series.c:
```diff
@@ -0,0 +1,2077 @@
+							enum gpio_pca_series_reg_type reg_type)
+{
+	const struct gpio_pca_series_config *cfg = dev->config;
+	/* struct gpio_pca_series_data *data = dev->data; */
```
Drop this line

danieldegrasse commented on drivers/gpio/gpio_pca_series.c:
```diff
@@ -0,0 +1,2077 @@
+	switch (reg_type) {
+	case gpio_pca_series_1b_input_port:
+	case gpio_pca_series_1b_output_port:
+/*  case gpio_pca_series_1b_polarity_inversion: */
```
```suggestion
```

danieldegrasse commented on drivers/gpio/gpio_pca_series.c:
```diff
@@ -0,0 +1,2077 @@
+					   enum gpio_pca_series_reg_type reg_type, uint8_t *buf)
+{
+	const struct gpio_pca_series_config *cfg = dev->config;
+	/* struct gpio_pca_series_data *data = dev->data; */
```
```suggestion
```

danieldegrasse commented on drivers/gpio/gpio_pca_series.c:
```diff
@@ -0,0 +1,2077 @@
+	msg[1].len = size;
+	msg[1].flags = I2C_MSG_RESTART | I2C_MSG_READ | I2C_MSG_STOP;
+
+	ret =  i2c_transfer(dev, msg, 2, addr);
```
Seems like this could be replaced with a call to `i2c_reg_read_byte`?

CkovMk replied:
Not really. device could have multiple bytes in one "register", stored `size`, either because each pin requires 2 bits of configuration, or the device has more than 1 port, or both.
In #69330 aasinclair stated:
> Burst write is unreliable, so please use i2c_write instead (See https://github.com/zephyrproject-rtos/zephyr/issues/20154).

So I fallbacked to these code.

danieldegrasse commented on drivers/gpio/gpio_pca_series.c:
```diff
@@ -0,0 +1,2077 @@
+					enum gpio_pca_series_reg_type reg_type, uint8_t *buf)
+{
+	const struct gpio_pca_series_config *cfg = dev->config;
+	/* struct gpio_pca_series_data *data = dev->data; */
```
```suggestion
```

danieldegrasse commented on drivers/gpio/gpio_pca_series.c:
```diff
@@ -0,0 +1,2077 @@
+	msg[1].len = size;
+	msg[1].flags = I2C_MSG_WRITE | I2C_MSG_STOP;
+
+	ret =  i2c_transfer(dev, msg, 2, addr);
```
Replace this block with `i2c_reg_write_byte`

danieldegrasse commented on drivers/gpio/gpio_pca_series.c:
```diff
@@ -0,0 +1,2077 @@
+	msg[1].len = size;
+	msg[1].flags = I2C_MSG_RESTART | I2C_MSG_READ | I2C_MSG_STOP;
+
+	ret =  i2c_transfer(dev, msg, 2, addr);
```
Replace with `i2c_reg_read_byte`

danieldegrasse commented on drivers/gpio/gpio_pca_series.c:
```diff
@@ -0,0 +1,2077 @@
+ * @{
+ */
+
+int gpio_pca_series_reset(const struct device *dev)
```
It doesn't seem like this function is used?

CkovMk replied:
it is used to 
1) assert reset pin if available
2) fallback to write reset value if reset pin is not configured.

I've implemented it and called it during init.

danieldegrasse commented on drivers/gpio/gpio_pca_series.c:
```diff
@@ -0,0 +1,2077 @@
+	return -ENOTSUP;
+}
+
+void gpio_pca_series_dump(const struct device *dev)
```
Does not appear this function is used- I understand needing it for debugging, but perhaps we could add somewhere where the function is called if `CONFIG_GPIO_LOG_LEVEL_DBG` is set? Otherwise, this is effectively dead code

danieldegrasse replied:
This still needs to be resolved

CkovMk replied:
Added KConfig option to dump all registers after init.

danieldegrasse commented on drivers/gpio/gpio_pca_series.c:
```diff
@@ -0,0 +1,2077 @@
+
+	k_sem_take(&data->lock, K_FOREVER);
+
+	/* TODO: Only write 1 byte */
```
Can you clarify what this comment means?

CkovMk replied:
Sure.
The config function only configures 1 pin, thus only 1 byte is changed per "register type". By only read-modify-write the changed byte, I2C transaction time can be optimized.

danieldegrasse replied:
Ah, I understand. We are writing all bytes of the register, but we only *need* to write one? Maybe add that information to the comment here, for future developers. This way the details of the optimization possible here aren't lost in the future

CkovMk replied:
I was planning to get it done in this pr. The problem here is that the current reg access api can only accept full reg (aka multi-byte). There has to be another api to modify a single byte.

CkovMk replied:
@danieldegrasse I'll leave a more detailed comments here, maybe try to optimize it in another pr, together with the drive strength api change.

This pr is large enough, I think I should focus on basic function and testing for now.

danieldegrasse commented on drivers/gpio/gpio_pca_series.c:
```diff
@@ -0,0 +1,2077 @@
+static int gpio_pca_series_port_read_extended(
+	const struct device *dev, gpio_port_value_t *value)
+{
+	/* const struct gpio_pca_series_config *cfg = dev->config; */
```
```suggestion
```

CkovMk commented:
Thanks. I'm out of office right now, will be back handling the comments in a week or so.

CkovMk commented on drivers/gpio/gpio_pca_series.c:
```diff
@@ -0,0 +1,2077 @@
+static const uint8_t gpio_pca_series_cache_map_pca953x[] = {
+	0x00, /** input_port, cache for interrupt */
+	0x01, /** output_port */
+/*  0x02, /** polarity_inversion */
```
The actual device has this register for inverting the output. Zephyr inverts output in software, so this register cannot be utilized. So in this driver it is not used. It is here just for completeness, so is the commented out line in enum `gpio_pca_series_reg_type`.

danieldegrasse replied:
Okay- so these registers just aren't used. Perhaps change the comment to something like so:

/* 0x02, polarity_inversion (unused, omitted from cache map) */

CkovMk replied:
sure.

CkovMk commented on drivers/gpio/gpio_pca_series.c:
```diff
@@ -0,0 +1,2077 @@
+	ARG_UNUSED(dev);
+	ARG_UNUSED(pins);
+
+	LOG_INF("gpio_int trigger");
```
debug purpose. will drop to DBG or remove it later.

CkovMk commented on boards/nxp/imx93_evk/imx93_evk_mimx9352_a55.dts:
```diff
@@ -122,6 +122,30 @@
 	};
+
+	gpio_exp1: pcal6524@22 {
+		compatible = "nxp,pca-series";
```
update this.

CkovMk replied:
updated.

CkovMk commented on drivers/gpio/Kconfig.pca_series:
```diff
@@ -0,0 +1,43 @@
+	help
+	  Enable driver for PCA_SERIES I2C-based GPIO chip.
+
+# TODO: add this line
```
add back this line when it is ready to retire pcal64xxa driver. currently they are conflict

CkovMk replied:
@danieldegrasse Is it better if I remove support for PCAL6408A and PCAL6416A in this pr, and open another pr to add back the support? Since retiring these drivers will need to edit a bunch of dts from other vendor...

CkovMk commented on dts/bindings/gpio/nxp,pca_series.yaml:
```diff
@@ -0,0 +1,48 @@
+# Copyright 2024 NXP
```
need to edit `nxp,pcal6408a.yaml` and `nxp,pcal6416a.yaml` when ready to retire those.

CkovMk commented on dts/bindings/gpio/nxp,pca9538.yaml:
```diff
@@ -0,0 +1,5 @@
+# SPDX-License-Identifier: Apache-2.0
+description: NXP PCA9538 I2C GPIO expander node
+compatible: nxp,pca9538
+include: nxp,pca_series.yaml
```
maybe I should add "ngpios" const property here, instead of hard-coding it in driver (macro `GPIO_PCA_GET_PORT_NO_BY_PART_NO`)? since now we are creating separate bindings for each device.
for example:

```yaml
properties:
  ngpios:
    required: true
    const: 8
```

Note that port_no is also hard-coded into each `struct gpio_pca_series_part_config` instance. I probably should remove it there, since maintaining port_no or pin_count at multiple places is inconvinient and error-prone.

CkovMk replied:
optimized.

CkovMk commented on drivers/gpio/gpio_pca_series.c:
```diff
@@ -0,0 +1,2077 @@
+	/** get current interrupt config */
+#ifdef CONFIG_GPIO_PCA_SERIES_CACHE_ALL
+	uint32_t buf[2];
+	/* FIXME: set all 0xff during init */
```
during init, cache should be set to proper state already.
I'll double check that, test on real hw to see if there are any bug.

CkovMk replied:
ckecked, not causing issue.

CkovMk commented on drivers/gpio/gpio_pca_series.c:
```diff
@@ -0,0 +1,2204 @@
+ * @return int 0 if success
+ *             -EIO if i2c error
+ */
+static int gpio_pca_series_pin_config_drive_strength(
```
This one is not used as well, as zephyr does not provide api for that. Not sure if it should be kept.

danieldegrasse replied:
No, we should drop code that will not be used by Zephyr

CkovMk replied:
Setting drive strength is common feature to gpio controllers. For SoC buit-in gpio controllers, this is usually set by pinctrl driver. But for gpio expander this does not apply. I went through issues and found #17072, which makes it possible to expand driver-specific api. Is it possible to expand gpio api with drive strength like that?

CkovMk replied:
sure, let's leave it to another pr.

CkovMk replied:
removed in this pr.

dleach02 commented:
@CkovMk please address the CI issues

danieldegrasse commented on drivers/gpio/gpio_pca_series.c:
```diff
@@ -0,0 +1,2178 @@
+
+/**
+ * }
+ * // gpio_pca_reg_access_api
```
Nit- please drop ``//`` style comments within this file- I think the CI check is not catching them since they are part of a C89 style comment block, but they should still be removed.

danieldegrasse commented on drivers/gpio/gpio_pca_series.c:
```diff
@@ -0,0 +1,2178 @@
+	if (mode == GPIO_INT_MODE_LEVEL) {
+		return -ENOTSUP;
+	}
+#ifdef DEBUG
```
I think we should probably create some form of Kconfig to control this behavior- We could also call the `gpio_pca_series_dump` function here, if we want to keep it present

CkovMk replied:
These code are meant for future developers: Checking feature flags in "extended" variant is only needed when developer wants to add new compatible device to this driver. For normal user, the api struct should've guaranteed that these flags are set, thus these code should be invisible to them. So it might be confusing if special KConfig is applied -- they are not supposed to see such err.

Btw, for completeness, I intended to add more checks in similar form. Please shed some light on how to properly handle this.

danieldegrasse replied:
I think adding a Kconfig to enable developer debugging options here makes sense, right? This way, developers can turn it on when they are working on the driver, but the Kconfig defaults to disabled, so a user won't normally see the error

danieldegrasse replied:
In Linux, it seems that driver-specific macros (which would not be exposed via Kconfig) are used: https://github.com/torvalds/linux/blob/98b1cc82c4affc16f5598d4fa14b1858671b2263/drivers/video/fbdev/cirrusfb.c#L1262. So I'm ok if we do something like this here, like `#ifdef NXP_PCA_DEBUG`

CkovMk replied:
> In Linux, it seems that driver-specific macros (which would not be exposed via Kconfig) are used: https://github.com/torvalds/linux/blob/98b1cc82c4affc16f5598d4fa14b1858671b2263/drivers/video/fbdev/cirrusfb.c#L1262. So I'm ok if we do something like this here, like #ifdef NXP_PCA_DEBUG

Then let's go this way : )

danieldegrasse commented on drivers/gpio/Kconfig.pca_series:
```diff
@@ -0,0 +1,50 @@
+	  On devices w/o interrupt status, it will also cache
+	  input state.
+
+config GPIO_PCA_SERIES_REGISTER_DUMP
```
Sorry that my prior comment wasn't clear- we can drop this Kconfig, or we can leave it in and make the code for register dumps depend on `CONFIG_GPIO_PCA_SERIES_REGISTER_DUMP`. Currently it depends on `GPIO_PCA_SERIES_REGISTER_DUMP`. Either method is ok with me, I just think we should be consistent

CkovMk replied:
Removed. Now `GPIO_NXP_PCA_SERIES_DEBUG` controls both register dump and extra checking.

CkovMk commented on drivers/gpio/gpio_pca_series.c:
```diff
@@ -0,0 +1,2169 @@
+		return -EFAULT;
+	}
+
+	ret = i2c_burst_read_dt(&cfg->i2c, addr, buf, size);
+	if (ret) {
+		LOG_ERR("i2c read error [%d]", ret);
+		return ret;
+	}
```
Switched back to burst_read api **temporarily**, because using `i2c_transfer` directly is causing i2c transaction to hang. Not sure why, will debug later.

CkovMk commented on drivers/gpio/gpio_pca_series.c:
```diff
@@ -0,0 +1,2169 @@
+		return -EFAULT;
+	}
+
+	ret = i2c_burst_write_dt(&cfg->i2c, addr, buf, size);
+	if (ret) {
+		LOG_ERR("i2c write error [%d]", ret);
+		return ret;
+	}
```
Temp fix, as stated before

CkovMk commented:
Tested with PCAL6524 on i.MX 93 EVK: `configure`, `read`, `write` apis are working. will test interrupt and full_cache mode.

CkovMk commented on drivers/gpio/gpio_pca_series.c:
```diff
@@ -0,0 +1,2169 @@
+#ifndef CONFIG_GPIO_PCA_SERIES_CACHE_ALL
+	/** cache input for interrupt handling */
+	gpio_pca_series_reg_cache_mini_get(dev)->input = input_data;
+	/* FIXME: handle interrupt here, fire callbacks as needed */
```
Need to address this.
For device w/o `PCA_HAS_INT_EXTEND` feature, an gpio read also clears interrupt flag. So after reading, driver needs to check if interrupt is triggered and fire callbacks.

danieldegrasse replied:
Based on this comment I assume you still  have some debug work planned before this PR is ready to merge. Could you do one of the following, so I know when this PR is ready for a final review?
- mark this PR as draft, and move it out of draft status when it is ready
- ping me explicitly when the PR is tested and ready to merge from your point of view

CkovMk replied:
sure. should be ready next week.

dleach02 replied:
be aware that Zephyr's next release has an RC1 feature freeze at the end of next week. You should strive to have this ready before this if you want this part of the next release

CkovMk replied:
Uhh...Just a update here...

I got some other work to do and just came back. I'm still working on this pr. I tested PCAL6524 on i.MX93 EVK, it's working fine w/o cache enabled. input, output, interrupt tested with `sample/basic/button` app.

Cache API, a little bit complicated...

Since PCAL6524 has all the registers, I'll be able to simulate other devices. So still need more testing.

CkovMk commented:
@danieldegrasse The code is in ready state.

I stripped away plenty of code and only left code supporting pcal6524 and pcal6534. I'll open new pr to expand it in the (near) future.

It is fully tested on i.MX 93 EVK with pcal6524 on board. A total of 4 cases tested: interrupt support enabled / disabled combined with cache all enabled and disabled.

- gpio shell is working properly.
- samples/basic/button working with interrupt enabled.

CkovMk committed:
boards: imx93_evk_mimx9352_a55: enable pcal6524 gpio expander

Enable pcal6524 gpio expander on i.MX 93 EVK

Signed-off-by: Chekhov Ma <chekhov.ma@nxp.com>

CkovMk commented:
ci issues resolved

CkovMk commented:
Soft ping to @danieldegrasse , Have you got time to review this? Thanks

danieldegrasse commented on drivers/gpio/gpio_pca_series.c:
```diff
@@ -0,0 +1,1727 @@
+		LOG_ERR("minimum cache failed to read initial input: %d", ret);
+	}
+#else
+	UNUSED(cfg);
```
I believe we should be able to replace these calls to `UNUSED` with `ARG_UNUSED`, and remove the `UNUSED` macro

CkovMk replied:
yes.

danieldegrasse commented on drivers/gpio/gpio_pca_series.c:
```diff
@@ -0,0 +1,1727 @@
+/** Private debug macro, enable more error checking and print more log. */
+/* #define GPIO_NXP_PCA_SERIES_DEBUG */
+
+#define UNUSED(x)           (void)(x)
```
We can drop this and use `ARG_UNUSED`, right?

CkovMk replied:
dropped.

danieldegrasse commented on drivers/gpio/gpio_pca_series.c:
```diff
@@ -0,0 +1,1727 @@
+		LOG_WRN("pca interrupt enabled w/o int-gpios configured in dts");
+	}
+#else
+	UNUSED(data);
```
For these and other `UNUSED` instances, could we simply not declare the unused variable unless an `#if` condition that would use it is true?

CkovMk replied:
Yeah, but that would add a lot of `#ifdef .. #endif` just to rule out an definition. I'd say I prefer the current solution.

CkovMk commented on drivers/gpio/gpio_pca_series.c:
```diff
@@ -0,0 +1,1727 @@
+	msg[1].buf = (uint8_t *)buf;
+	msg[1].len = size;
+	msg[1].flags = I2C_MSG_WRITE | I2C_MSG_STOP;
+	ret = i2c_transfer_dt(&cfg->i2c, msg, 2);
```
i2c_write requires 1 single buffer, but we have reg_addr seperated from the data buffer. So splitting into 2 requests saves 1 copy.

CkovMk replied:
I just noticed #38678, which seems to fix #20154. I assume it's safe to use "concat_write" here. I'm still not sure the future to i2c_burst apis and i2c_write_read api, so let's just use the i2c_transfer api for now.

CkovMk replied:
add @aasinclair , I wonder if you know the roadmap / known issues of current I2C apis?

danieldegrasse replied:
Looking at this more you are right- there isn't really a good I2C helper function here. Let's just keep this as is.

CkovMk commented on drivers/gpio/gpio_pca_series.c:
```diff
@@ -0,0 +1,1727 @@
+/**
+ * @brief part number definition.
+ *
+ * @note must be consistent with @ref part_no in dt binding
```
It's effectivly saying "the order must be the same". I'll reword it to "@note must be consistent in order with @ref part_no in dt binding"

CkovMk commented on drivers/gpio/gpio_pca_series.c:
```diff
@@ -0,0 +1,1727 @@
+	msg[1].buf = (uint8_t *)buf;
+	msg[1].len = size;
+	msg[1].flags = I2C_MSG_RESTART | I2C_MSG_READ | I2C_MSG_STOP;
+	ret = i2c_transfer_dt(&cfg->i2c, msg, 2);
```
It's the same code as `i2c_write_read`. I chose to expand it just to maintain symmetry to the below write function. It's ok for me to change it.

danieldegrasse replied:
Yeah, let's avoid duplicating the implementation

CkovMk replied:
changed.

CkovMk commented:
also add pca9554 and 9555, which are basically variants of pca9538 and pca9539 respectively.

CkovMk committed:
drivers: gpio: pca_series: add api to get per-port register size

This commit add get per-port reg size api to make it more clear to
understand the design.

Signed-off-by: Chekhov Ma <chekhov.ma@nxp.com>

CkovMk commented:
fixed bug when configuring pin input/output.

CkovMk committed:
drivers: gpio: pca_series: add supported device to tests/build_all

Add pcal6524, pcal6534, pca9538, pca9539 to tests/build_all/gpio

Signed-off-by: Chekhov Ma <chekhov.ma@nxp.com>

CkovMk commented:
uh there's an issue when multiple part# is present in the same app. inst# is reused, causing redefinition of inst##_cfg and inst##_data. need to encode part no into variable names.

haven't anticipate that...

CkovMk commented:
above issue is fixed.

Now CI is failing because

```
zephyr/drivers/gpio/gpio_pca_series.c:524:79: error: cast from pointer to integer of different size [-Werror=pointer-to-int-cast]
  524 |         LOG_DBG("cache read type %d len %d mem addr 0x%8.8x", reg_type, size, (uint32_t)src);
```

If I use `%lx` and `(size_t)ptr` conversion, it will report error on 32bit machines. If I use `%x` with `(uint32_t)ptr` conversion, it will report error on 64bit machines.

Not sure what to do...

henrikbrixandersen commented:
> above issue is fixed.
> 
> Now CI is failing because
> 
> ```
> zephyr/drivers/gpio/gpio_pca_series.c:524:79: error: cast from pointer to integer of different size [-Werror=pointer-to-int-cast]
>   524 |         LOG_DBG("cache read type %d len %d mem addr 0x%8.8x", reg_type, size, (uint32_t)src);
> ```
> 
> If I use `%lx` and `(size_t)ptr` conversion, it will report error on 32bit machines. If I use `%x` with `(uint32_t)ptr` conversion, it will report error on 64bit machines.
> 
> Not sure what to do...

Pointers can be printed using the `%p` format string.

CkovMk commented:
> Pointers can be printed using the %p format string.

Oh, one nice tip : )

CkovMk committed:
drivers: gpio: add pca_series gpio expander driver

There are numbers of drivers for different PCA(L) series chip. They
share similiar register layout and control logic. This driver intends
to unify these drivers for PCA(L)xxxx series i2c gpio expanders.

Signed-off-by: Chekhov Ma <chekhov.ma@nxp.com>

CkovMk committed:
drivers: gpio: extend pca_series driver to pca953x and pca955x

This commit extends pca_series gpio driver to devices pca9538, pca9539,
pca9554 and pca9555.

Signed-off-by: Chekhov Ma <chekhov.ma@nxp.com>

CkovMk committed:
drivers: gpio: pca_series: handle byteorder

Add byteorder handling to support big-endian machine.

Signed-off-by: Chekhov Ma <chekhov.ma@nxp.com>

cyberjupiter commented:
Hi, thank you for the driver. I know this is merged but just want ask is it possible to include samples on how to use different pca9xxx to configure a single input/output pin? (not an entire port).

CkovMk commented:
@cyberjupiter This driver can be used just like other GPIO drivers, input / output is configured via Zephyr GPIO API. You may refer to `samples/basic/button`.

What else would you like to know with a new sample?

CkovMk commented:
@cyberjupiter It depends on board design. For i.MX 93 EVK it's possible to connect 2 onboard buttons to PCAL6524 chip, by flipping a series of dip switches. I could provice such use casse as a dt overlay, is that something u r intersted in?

cyberjupiter commented:
@CkovMk Yes, that would be great. TIA.

CkovMk commented:
@cyberjupiter you may refer to #78693
